File: src/security/webhook.ts
// Webhook / inbound-event signature verification.
//
// Every event that enters the recovery pipeline from an external system (e.g.
// a Razorpay `subscription.charged.failed` webhook, checkout analytics, an ERP
// receivable extract) must be authenticated BEFORE it is trusted as an input.
//
// Razorpay signs webhook payloads with a scheme that carries both a timestamp
// and an HMAC-SHA256 signature:
//
//   Header:  X-Razorpay-Signature: t=<unix_seconds>|s=<hex_hmac_sha256>
//   Payload: HMAC_SHA256(secret, `${t}.${rawBody}`)
//
// We verify the signature AND check the timestamp is fresh (replay protection).
// Nothing downstream runs until verification passes.

import { createHmac, timingSafeEqual } from 'crypto';

export interface VerifiedWebhook {
  ok: boolean;
  eventId: string;
  flow: string;
  reason?: string; // when !ok
}

// Max age (seconds) a signed payload may be before we reject it as a replay.
export const MAX_WEBHOOK_SKEW_SECONDS = 300; // 5 min

function parseSignatureHeader(header: string | undefined): { t: number; s: string } | null {
  if (!header) return null;
  const parts = header.split('|').reduce<Record<string, string>>((acc, kv) => {
    const idx = kv.indexOf('=');
    if (idx > 0) acc[kv.slice(0, idx)] = kv.slice(idx + 1);
    return acc;
  }, {});
  const t = Number(parts['t']);
  const s = parts['s'];
  if (!Number.isFinite(t) || !s) return null;
  return { t, s };
}

// Constant-time compare to avoid timing side-channels.
function safeEqual(a: string, b: string): boolean {
  const ab = Buffer.from(a, 'utf8');
  const bb = Buffer.from(b, 'utf8');
  if (ab.length !== bb.length) return false;
  return timingSafeEqual(ab, bb);
}

// Compute the Razorpay-style signature for a payload + timestamp.
export function computeWebhookSignature(secret: string, t: number, rawBody: string): string {
  return createHmac('sha256', secret).update(`${t}.${rawBody}`).digest('hex');
}

// Verify a webhook. `secret` comes from config/env (see secrets.ts) — never a
// hard-coded value. `nowSeconds` injectable for deterministic tests.
export function verifyWebhook(
  secret: string,
  headers: Record<string, string | undefined>,
  rawBody: string,
  nowSeconds?: number,
): { ok: boolean; reason?: string } {
  const sig = parseSignatureHeader(
    headers['x-razorpay-signature'] ?? headers['X-Razorpay-Signature'],
  );
  if (!sig) return { ok: false, reason: 'missing_or_malformed_signature' };

  const now = nowSeconds ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - sig.t) > MAX_WEBHOOK_SKEW_SECONDS) {
    return { ok: false, reason: 'expired_signature_replay_rejected' };
  }

  const expected = computeWebhookSignature(secret, sig.t, rawBody);
  if (!safeEqual(expected, sig.s)) {
    return { ok: false, reason: 'signature_mismatch' };
  }
  return { ok: true };
}

// Convenience: verify the body, then extract the event id/flow from the trusted
// payload. Never call this with an unverified body; the classifer/batch must
// only ever see events that passed verifyWebhook.
export function parseTrustedEvent(rawBody: string): { eventId: string; flow?: string } | null {
  try {
    const parsed = JSON.parse(rawBody);
    const eventId: string | undefined = parsed?.event_id ?? parsed?.id;
    const flow: string | undefined = parsed?.flow ?? parsed?.entity?.type;
    if (!eventId) return null;
    return { eventId, flow };
  } catch {
    return null;
  }
}

File: go/rz/webhook.go
package rz

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"strings"
	"time"
)

// MAX_WEBHOOK_SKEW_SECONDS is the replay window.
const MAX_WEBHOOK_SKEW_SECONDS = 300

type VerifiedWebhook struct {
	OK      bool
	EventId string
	Flow    string
	Reason  string
}

func parseSignatureHeader(header string) (t int64, s string, ok bool) {
	if header == "" {
		return
	}
	pairs := strings.Split(header, "|")
	m := map[string]string{}
	for _, kv := range pairs {
		i := strings.IndexByte(kv, '=')
		if i > 0 {
			m[kv[:i]] = kv[i+1:]
		}
	}
	if m["t"] == "" || m["s"] == "" {
		return
	}
	var v int64
	for _, c := range m["t"] {
		if c < '0' || c > '9' {
			return
		}
		v = v*10 + int64(c-'0')
	}
	return v, m["s"], true
}

func safeEqual(a, b string) bool {
	if len(a) != len(b) {
		return false
	}
	ab := []byte(a)
	bb := []byte(b)
	return hmac.Equal(ab, bb)
}

// ComputeWebhookSignature produces the Razorpay-style t=<unix>|s=<hex_hmac> value.
func ComputeWebhookSignature(secret string, t int64, rawBody string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(fmt.Sprintf("%d.%s", t, rawBody)))
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifyWebhook checks freshness + HMAC. Secret MUST come from config/env
// (see secrets.go), never a hard-coded value. nowSeconds injectable for
// deterministic tests.
func VerifyWebhook(secret string, headers map[string]string, rawBody string, nowSeconds int64) (ok bool, reason string) {
	hdr := headers["x-razorpay-signature"]
	if hdr == "" {
		hdr = headers["X-Razorpay-Signature"]
	}
	t, sig, parsed := parseSignatureHeader(hdr)
	if !parsed {
		return false, "missing_or_malformed_signature"
	}

	now := nowSeconds
	if now == 0 {
		now = time.Now().Unix()
	}
	if math.Abs(float64(now-t)) > MAX_WEBHOOK_SKEW_SECONDS {
		return false, "expired_signature_replay_rejected"
	}

	expected := ComputeWebhookSignature(secret, t, rawBody)
	if !safeEqual(expected, sig) {
		return false, "signature_mismatch"
	}
	return true, ""
}

// ParseTrustedEvent extracts eventId/flow from an ALREADY VERIFIED body.
func ParseTrustedEvent(rawBody string) (eventId string, flow string, ok bool) {
	if strings.TrimSpace(rawBody) == "" {
		return "", "", false
	}
	var payload struct {
		EventID string `json:"event_id"`
		ID      string `json:"id"`
		Flow    string `json:"flow"`
		Entity  struct {
			Type string `json:"type"`
		} `json:"entity"`
	}
	if err := json.Unmarshal([]byte(rawBody), &payload); err != nil {
		return "", "", false
	}
	eventId = payload.EventID
	if eventId == "" {
		eventId = payload.ID
	}
	flow = payload.Flow
	if flow == "" {
		flow = payload.Entity.Type
	}
	if eventId == "" {
		return "", "", false
	}
	return eventId, flow, true
}
